packages/panzer/disc-fe/src/Panzer_BasisValues2_impl.hpp

```diff
@@ -45,1711 +45,1628 @@
+  auto host_orientations = Kokkos::create_mirror_view(device_orientations);
+  for(int i=0; i < num_cells; ++i)
+    host_orientations(i) = orientations[i];
+  Kokkos::deep_copy(device_orientations,host_orientations);
+
+  if (elmtspace == PureBasis::HGRAD){
+    applyOrientationsImpl<Scalar>(num_cells, basis_scalar.get_view(), device_orientations, *intrepid_basis);
+    if(build_weighted) applyOrientationsImpl<Scalar>(num_cells, weighted_basis_scalar.get_view(), device_orientations, *intrepid_basis);
+    if(compute_derivatives) applyOrientationsImpl<Scalar>(num_cells, grad_basis.get_view(), device_orientations, *intrepid_basis);
+    if(compute_derivatives and build_weighted) applyOrientationsImpl<Scalar>(num_cells, weighted_grad_basis.get_view(), device_orientations, *intrepid_basis);
+  }
+
+  if(elmtspace == PureBasis::HCURL){
+    applyOrientationsImpl<Scalar>(num_cells, basis_vector.get_view(), device_orientations, *intrepid_basis);
+    if(build_weighted) applyOrientationsImpl<Scalar>(num_cells, weighted_basis_vector.get_view(), device_orientations, *intrepid_basis);
+    if(num_dim == 2){
+      if(compute_derivatives) applyOrientationsImpl<Scalar>(num_cells, curl_basis_scalar.get_view(), device_orientations, *intrepid_basis);
+      if(compute_derivatives and build_weighted) applyOrientationsImpl<Scalar>(num_cells, weighted_curl_basis_scalar.get_view(), device_orientations, *intrepid_basis);
+    }
+    if(num_dim == 3){
+      if(compute_derivatives) applyOrientationsImpl<Scalar>(num_cells, curl_basis_vector.get_view(), device_orientations, *intrepid_basis);
+      if(compute_derivatives and build_weighted) applyOrientationsImpl<Scalar>(num_cells, weighted_curl_basis_vector.get_view(), device_orientations, *intrepid_basis);
+    }
+  }
+
+  if(elmtspace == PureBasis::HDIV){
+    applyOrientationsImpl<Scalar>(num_cells, basis_vector.get_view(), device_orientations, *intrepid_basis);
+    if(build_weighted) applyOrientationsImpl<Scalar>(num_cells, weighted_basis_vector.get_view(), device_orientations, *intrepid_basis);
+    if(compute_derivatives) applyOrientationsImpl<Scalar>(num_cells, div_basis.get_view(), device_orientations, *intrepid_basis);
+    if(compute_derivatives and build_weighted) applyOrientationsImpl<Scalar>(num_cells, weighted_div_basis.get_view(), device_orientations, *intrepid_basis);
+  }
+
+  orientations_applied_ = true;
+}
+
+// method for applying orientations
+template <typename Scalar>
+void BasisValues2<Scalar>::
+applyOrientations(const PHX::MDField<const Scalar,Cell,BASIS> & orientations)
+{
+  TEUCHOS_TEST_FOR_EXCEPT_MSG(true,"panzer::BasisValues2::applyOrientations : this should not be called.");
+}
+
+template <typename Scalar>
+PureBasis::EElementSpace BasisValues2<Scalar>::getElementSpace() const
+{ return basis_layout->getBasis()->getElementSpace(); }
+
+template <typename Scalar>
+void panzer::BasisValues2<Scalar>::
+setupArrays(const Teuchos::RCP<const panzer::BasisIRLayout>& layout,
```

Comment:
Currently the only reason for this call (AFAICT) is the BasisValues_Evaluator in Panzer, which needs to register existing MDFields from the BasisValues2 object with the FieldManager. That one pattern leads to some very complex behavior behind the scenes in BasisValues2 when caching data (e.g. we can't just overwrite the existing data).

Supposedly, if we don't pre-allocate the arrays we can do the lazy evaluate just fine and save memory. Otherwise, yeah... all arrays are allocated from the start.